Sources/FluentKit/Properties/Parent+EagerLoad.swift

```diff
@@ -12,101 +12,105 @@ extension Parent: AnyEagerLoadable {
+        if let subquery = request as? SubqueryEagerLoad {
             self.eagerLoadedValue = try subquery.get(id: id)
         } else {
             fatalError("unsupported eagerload request: \(request)")
         }
     }
 }
 
+
+// MARK: - Eager Loadable
+
 extension Parent: EagerLoadable {
-    public func eagerLoad<Model>(to builder: QueryBuilder<Model>)
-        where Model: FluentKit.Model
-    {
-        builder.eagerLoads.requests[self.eagerLoadKey] = ParentSubqueryEagerLoad<To>(
-            key: self.$id.key
-        )
+    public func eagerLoad<Model>(to builder: QueryBuilder<Model>) where Model: FluentKit.Model {
+        builder.eagerLoads.requests[self.eagerLoadKey] = self.eagerLoadRequest
     }
 }
 
+// MARK: Private
 
-extension OptionalParent: AnyEagerLoadable {
-    var eagerLoadKey: String {
-        return "p:" + self.$id.key
-    }
+extension Parent {
+    internal final class SubqueryEagerLoad: EagerLoadRequest {
+        typealias Loader = (Database, Set<To.IDValue>) -> EventLoopFuture<[To]>
+        typealias LoadedParent = (id: To.IDValue, model: Parent<To>.EagerLoaded)
 
-    var eagerLoadValueDescription: CustomStringConvertible? {
-        return self.eagerLoadedValue
-    }
+        let key: String
+        let loader: Loader
 
-    func eagerLoad(from eagerLoads: EagerLoads) throws {
-        guard let request = eagerLoads.requests[self.eagerLoadKey] else {
-            return
+        var storage: [LoadedParent]
+
+        var description: String {
+            return self.storage.description
         }
 
-        self.didEagerLoad = true
-        guard let id = self.id else {
-            return
+        private init(key: String, loader: @escaping Loader) {
+            self.storage = []
+            self.loader = loader
+            self.key = key
         }
 
-        if let subquery = request as? ParentSubqueryEagerLoad<To> {
-            self.eagerLoadedValue = try subquery.get(id: id)
-        } else {
-            fatalError("unsupported eagerload request: \(request)")
+        func prepare(query: inout DatabaseQuery) {
+            // no preparation needed
         }
-    }
-}
 
-extension OptionalParent: EagerLoadable {
-    public func eagerLoad<Model>(to builder: QueryBuilder<Model>)
-        where Model: FluentKit.Model
-    {
-        builder.eagerLoads.requests[self.eagerLoadKey] = ParentSubqueryEagerLoad<To>(
-            key: self.$id.key
-        )
+        func run(models: [AnyModel], on database: Database) -> EventLoopFuture<Void> {
+            var storedNil = false
+            self.storage = models.compactMap { model -> LoadedParent? in
+                guard let parent = try! model.anyID.cachedOutput!.decode(self.key, as: To.IDValue?.self) else {
+                    guard !storedNil && _isOptional(To.self) else { return nil }
+                    storedNil = true
+
+                    return (Optional<Void>.none as! To.IDValue, .loaded(Optional<Void>.none as! To))
```

Comment:
What is this code doing?

Reply:
What we are doing here is pre-storing a value for each parent ID in the `.storage` array. For models that exist, that value we store is `.notLoaded`, until we replace it later when we run the actual database query. Now, if the `Parent.To` type is optional, the parent may not exist. We can check that before we run the query by looking to see if the parent ID is `nil`. If it is, we know that the parent does not exist, so we set the pre-loaded value to `.loaded(nil)`. I talked to Joannis about this, and he agreed that since all `nil` parent IDs will always result in a `nil` parent, we only have to load that `nil` value into the `.storage` array once and increase our performance, instead of adding a duplicate `nil` element for each empty parent.

Reply:
I'm concerned about how reliable this code will be. Is there any way we could protocolize this so that we don't have to use `as!`? Or could loaded storage just be an optional?